File: sample/target.swift
import Foundation

protocol Validatable {
    func validate(key: String) -> Result<Void, ValidationError>
}

enum ValidationError: Error, CustomStringConvertible {
    case wrongLength(got: Int, expected: Int)
    case checksumMismatch(got: Int, expected: Int)
    case wrongPrefix(got: Character, expected: Character)

    var description: String {
        switch self {
        case .wrongLength(let g, let e):      return "key length \(g), expected \(e)"
        case .checksumMismatch(let g, let e): return "checksum \(g), expected \(e)"
        case .wrongPrefix(let g, let e):      return "prefix '\(g)', expected '\(e)'"
        }
    }
}

// Generic struct — monomorphization produces distinct $s symbols per concrete type
struct DataProcessor<T: Numeric & CVarArg> {
    var data: [T]

    mutating func transform(multiplier: T, offset: T) {
        data = data.map { $0 * multiplier + offset }
    }
}

struct License {
    let key: String
    static let requiredLength   = 12
    static let requiredChecksum = 0x4B2

    func isValid() -> Result<Void, ValidationError> {
        guard key.count == Self.requiredLength else {
            return .failure(.wrongLength(got: key.count, expected: Self.requiredLength))
        }
        let sum = key.unicodeScalars.reduce(0) { $0 + Int($1.value) }
        guard sum == Self.requiredChecksum else {
            return .failure(.checksumMismatch(got: sum, expected: Self.requiredChecksum))
        }
        return .success(())
    }
}

final class LicenseValidator: Validatable {
    let name: String
    let requiredPrefix: Character

    init(name: String, requiredPrefix: Character) {
        self.name = name
        self.requiredPrefix = requiredPrefix
    }

    func validate(key: String) -> Result<Void, ValidationError> {
        guard key.first == requiredPrefix else {
            return .failure(.wrongPrefix(got: key.first ?? "?", expected: requiredPrefix))
        }
        return License(key: key).isValid()
    }
}

let args = CommandLine.arguments
guard args.count >= 2 else {
    fputs("Usage: \(args[0]) <license_key>\n", stderr)
    exit(1)
}

let validator = LicenseValidator(name: "ProdValidator", requiredPrefix: "A")

switch validator.validate(key: args[1]) {
case .success:
    print("Valid license!")
case .failure(let e):
    fputs("Invalid license: \(e)\n", stderr)
    exit(1)
}

var proc = DataProcessor(data: [1, 2, 3, 4, 5] as [Int32])
proc.transform(multiplier: 2, offset: 0xDEAD)
print("Processed: \(proc.data[0]) \(proc.data[1]) \(proc.data[2])")

File: sample/target.rs
use std::env;

trait Validate {
    fn validate(&self, key: &str) -> Result<(), String>;
}

// Generic struct — monomorphization produces distinct _R symbols per T
struct DataProcessor<T> {
    data: Vec<T>,
}

impl<T: Copy + std::ops::Mul<Output = T> + std::ops::Add<Output = T> + From<u16>>
    DataProcessor<T>
{
    fn new(data: Vec<T>) -> Self {
        DataProcessor { data }
    }

    fn transform(&mut self, multiplier: T, offset: T) {
        for x in self.data.iter_mut() {
            *x = *x * multiplier + offset;
        }
    }

    fn data(&self) -> &[T] {
        &self.data
    }
}

struct License<'a> {
    key: &'a str,
}

impl<'a> License<'a> {
    fn new(key: &'a str) -> Self {
        License { key }
    }

    fn is_valid(&self) -> bool {
        if self.key.len() != 12 {
            return false;
        }
        let sum: i32 = self.key.bytes().map(|b| b as i32).sum();
        sum == 0x4B2
    }
}

struct LicenseValidator {
    _name: String,
    required_prefix: char,
}

impl LicenseValidator {
    fn new(name: &str, required_prefix: char) -> Self {
        LicenseValidator {
            _name: name.to_string(),
            required_prefix,
        }
    }
}

impl Validate for LicenseValidator {
    fn validate(&self, key: &str) -> Result<(), String> {
        if !key.starts_with(self.required_prefix) {
            return Err(format!("key must start with '{}'", self.required_prefix));
        }
        let lic = License::new(key);
        if lic.is_valid() {
            Ok(())
        } else {
            Err("checksum mismatch".to_string())
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <license_key>", args[0]);
        std::process::exit(1);
    }

    let validator = LicenseValidator::new("ProdValidator", 'A');

    match validator.validate(&args[1]) {
        Ok(()) => println!("Valid license!"),
        Err(e) => {
            eprintln!("Invalid license: {e}");
            std::process::exit(1);
        }
    }

    let mut proc = DataProcessor::new(vec![1i32, 2, 3, 4, 5]);
    proc.transform(2, 0xDEAD);
    let d = proc.data();
    println!("Processed: {} {} {}", d[0], d[1], d[2]);
}
